docs/tutorials/intro-to-rust.md

```diff
@@ -0,0 +1,418 @@
+</blockquote>
+
+## Resources
+
+The online book [from the official Rust site](https://doc.rust-lang.org/1.30.0/book/2018-edition/ch00-00-introduction.html) is a great resource to start with. It's recommended to read Chapter 1 before continuing, especially the `cargo` section.
+
+## Getting started
+
+Some smart contract examples from NEAR have a primary file that holds the code: `/src/lib.rs`. This is the conventional filename for a Rust library. Libraries will work great for compiling into WebAssembly and deployed the blockchain.
+
+The example shown here will use a `lib.rs` file with smart contract logic using a `struct`, the `struct`'s functions, and unit tests. This will all be in one file for this simple example. As developers build more complex smart contracts, it's a good idea to [organize code the Rust way](https://doc.rust-lang.org/1.30.0/book/2018-edition/ch12-03-improving-error-handling-and-modularity.html#separation-of-concerns-for-binary-projects).
+
+We'll break down the code in pieces below. If you wish to preview the complete code, please expand the bullets below.
+
+## Full example files
+
+<details>
+  <summary>Full `Cargo.toml` file</summary>
+
+```rust
+[package]
+name = "rust-counter-tutorial"
+version = "0.1.0"
+authors = ["Alice Bob <alice@example.com>"]
+edition = "2018"
+
+[lib]
+crate-type = ["cdylib", "rlib"]
+
+[dependencies]
+serde = { version = "1.0", features = ["derive"] }
+serde_json = "1.0.45"
+near-bindgen = "0.6.0"
+borsh = "0.6.1"
+wee_alloc = "0.4.5"
+
+[profile.release]
+codegen-units = 1
+# Tell `rustc` to optimize for small code size.
+opt-level = "z"
+lto = true
+debug = false
+panic = "abort"
+
+[workspace]
+members = []
+```
+
+</details>
+
+<details>
+  <summary>Full `src/lib.rs` file</summary>
+  
+```rust
+use borsh::{BorshDeserialize, BorshSerialize};
+use serde::{Deserialize, Serialize};
+use near_bindgen::{env, near_bindgen};
+
+#[global_allocator]
+static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;
+
+// add the following attributes to prepare your code for serialization and invocation on the blockchain
+// More built-in Rust attributes here: https://doc.rust-lang.org/reference/attributes.html#built-in-attributes-index
+#[near_bindgen]
+#[derive(Default, BorshDeserialize, BorshSerialize)]
+#[derive(Serialize, Deserialize)]
+pub struct Counter {
+    val: i8, // i8 is signed. unsigned integers are also available: u8, u16, u32, u64, u128
+}
+
+#[near_bindgen]
+impl Counter {
+    // returns 8-bit signed integer, must match the type from our struct's 'val' defined above
+    pub fn get_num(&self) -> i8 {
+        return self.val;
+    }
+    
+    // increment the counter
+    pub fn increment(&mut self) {
+        // note: adding one like this is an easy way to accidentally overflow
+        // real smart contracts will want to have safety checks
+        self.val = self.val + 1;
+        let _log_message = format!("Increased number to {}", self.val);
```

Comment:
It doesn't need to be prefixed with `_` since it is used below, does it?

Reply:
beautiful. These kinds of comments are helpful, thank you. I don't spend my days writing Rust and welcome feedback like this. Honestly, I didn't know, so thanks, fixed.